File: Demo/src/main/java/com/paypal/demo/MainActivity.java
package com.paypal.demo;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.braintreepayments.api.exceptions.InvalidArgumentException;
import com.braintreepayments.api.models.CardBuilder;
import com.paypal.androidsdk.CheckoutClient;
import com.paypal.androidsdk.PayPalCheckoutListener;
import com.paypal.androidsdk.PayPalCheckoutResult;
import com.paypal.androidsdk.interfaces.CheckoutListener;
import com.paypal.androidsdk.models.CheckoutResult;
import com.paypal.demo.models.Amount;
import com.paypal.demo.models.Order;
import com.paypal.demo.models.OrderRequest;
import com.paypal.demo.models.PayPalUAT;
import com.paypal.demo.models.Payee;
import com.paypal.demo.models.PurchaseUnit;

import java.util.ArrayList;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity implements PayPalCheckoutListener {

    private CheckoutListener mCheckoutListener;
    private DemoAPIClient mDemoClient;
    private PayPalUAT mPayPalUAT;
    private String mOrderID;

    private TextView mStatusLabel;
    private TextView mUATLabel;
    private TextView mOrderIDLabel;

    private CheckoutClient checkoutClient;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mStatusLabel = findViewById(R.id.statusTextView);
        mUATLabel = findViewById(R.id.uatTextView);
        mOrderIDLabel = findViewById(R.id.orderIDTextView);

        mDemoClient = RetrofitClientInstance.getInstance().create(DemoAPIClient.class);

        setUpListeners();
        fetchUAT();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (checkoutClient != null) {
            checkoutClient.resume(this);
        }
    }

    private Exception createUATNotFoundError() {
        return new IllegalStateException(getString(R.string.error_uat_not_found));
    }

    private void fetchUAT() {
        mUATLabel.setText(R.string.uat_fetching);

        Call<PayPalUAT> call = mDemoClient.getPayPalUAT("US");
        call.enqueue(new Callback<PayPalUAT>() {
            @Override
            public void onResponse(Call<PayPalUAT> call, Response<PayPalUAT> response) {
                Exception error = null;
                String uatString = null;

                mPayPalUAT = response.body();
                if (mPayPalUAT == null) {
                    error = createUATNotFoundError();
                } else {
                    try {
                        uatString = mPayPalUAT.getUAT();
                        if (uatString == null) {
                            error = createUATNotFoundError();
                        } else {
                            checkoutClient = new CheckoutClient(uatString, MainActivity.this);
                        }
                    } catch (InvalidArgumentException e) {
                        error = e;
                    }
                }

                if (error != null) {
                    mUATLabel.setText(getString(R.string.uat_display, error.getMessage()));
                    error.printStackTrace();
                }

                if (uatString != null) {
                    mUATLabel.setText(getString(R.string.uat_display, uatString));
                }
            }

            @Override
            public void onFailure(Call<PayPalUAT> call, Throwable e) {
                mUATLabel.setText(getString(R.string.uat_display, e.getMessage()));
            }
        });
    }

    public void fetchOrderId(View view) {
        // Construct order request
        // TODO: make this more concise
        Amount amount = new Amount();
        amount.setCurrencyCode("USD");
        amount.setValue("10.00");

        Payee payee = new Payee();
        payee.setEmailAddress("ahuang-ppcp-demo-sb1@paypal.com");

        PurchaseUnit purchaseUnit = new PurchaseUnit();
        purchaseUnit.setAmount(amount);
        purchaseUnit.setPayee(payee);

        ArrayList<PurchaseUnit> purchaseUnits = new ArrayList<>();
        purchaseUnits.add(purchaseUnit);

        OrderRequest orderRequest = new OrderRequest();
        orderRequest.setIntent("CAPTURE");
        orderRequest.setPurchaesUnits(purchaseUnits);

        // Fetch order
        mOrderIDLabel.setText(R.string.order_id_fetching);
        Call<Order> call = mDemoClient.fetchOrderID("US", orderRequest);
        call.enqueue(new Callback<Order>() {
            @Override
            public void onResponse(@NonNull Call<Order> call, @NonNull Response<Order> response) {
                Order order = response.body();
                if (order != null) {
                    mOrderID = order.getID();
                }

                if (mOrderID == null) {
                    mOrderIDLabel.setText(getString(R.string.error_order_id_not_found));
                } else {
                    mOrderIDLabel.setText(getString(R.string.order_id_display, mOrderID));
                }
            }

            @Override
            public void onFailure(Call<Order> call, Throwable t) {
                mOrderIDLabel.setText(getString(R.string.order_id_error));
            }
        });
    }

    private void setUpListeners() {
        mCheckoutListener = new CheckoutListener() {
            @Override
            public void onResult(@Nullable Exception error, @Nullable CheckoutResult result) {
                String status = null;
                if (result != null) {
                    status = getString(R.string.checkout_success, result.getOrderID());
                } else if (error != null) {
                    status = getString(R.string.checkout_error, error.getMessage());
                }
                mStatusLabel.setText(status);
            }
        };
    }

    // payment handler checkout implementations

    public void initiateCardCheckout(View view) {
        mStatusLabel.setText(getString(R.string.checkout_initiated));

        // trigger 3ds v1
        CardBuilder cardBuilder = new CardBuilder()
                .cardholderName("Suzie Smith")
                .cardNumber("4000000000000002")
                .expirationMonth("01")
                .expirationYear("2023")
                .cvv("123");

//        CardBuilder cardBuilder = new CardBuilder()
//                .cardholderName("Suzie Smith")
//                .cardNumber("4111111111111111")
//                .expirationMonth("01")
//                .expirationYear("2023")
//                .cvv("123");

        checkoutClient.payWithCard(cardBuilder, mOrderID, this, mCheckoutListener);
    }

    public void initiatePayPalCheckout(View view) {
        checkoutClient.payWithPayPal(mOrderID, this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.settings:
                this.startActivity(new Intent(this, SettingsActivity.class));
                return true;
            case R.id.reset:
                fetchUAT();
                mOrderID = null;
                mOrderIDLabel.setText(getString(R.string.order_id_placeholder));
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onResult(@Nullable Exception e, @Nullable PayPalCheckoutResult result) {
        // TODO: process result
    }
}

File: Demo/src/main/java/com/paypal/demo/MainActivity.kt
package com.paypal.demo

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.braintreepayments.api.Card
import com.braintreepayments.api.models.CardBuilder
import com.braintreepayments.api.models.PaymentMethodNonce
import com.braintreepayments.cardform.view.ExpirationDateEditText
import com.paypal.androidsdk.*
import com.paypal.demo.models.Amount
import com.paypal.demo.models.OrderRequest
import com.paypal.demo.models.Payee
import com.paypal.demo.models.PurchaseUnit

class MainActivity : AppCompatActivity(), CheckoutListener {

    private val merchant = Merchant()
    private var checkoutClient: CheckoutClient? = null

    private var orderId: String? = null

    private val uatLabel: TextView by lazy { findViewById<TextView>(R.id.uatTextView) }
    private val statusLabel: TextView by lazy { findViewById<TextView>(R.id.statusTextView) }
    private val orderIdLabel: TextView by lazy { findViewById<TextView>(R.id.orderIDTextView) }

    private val cardNumberField: EditText by lazy { findViewById<EditText>(R.id.bt_card_form_card_number) }
    private val expDateField: ExpirationDateEditText by lazy { findViewById<ExpirationDateEditText>(R.id.bt_card_form_expiration) }
    private val cvvField: EditText by lazy { findViewById<EditText>(R.id.bt_card_form_cvv) }

    private val progressBar: ProgressBar by lazy { findViewById<ProgressBar>(R.id.loading_spinner)}

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        fetchUAT()
    }

    override fun onResume() {
        super.onResume()
        checkoutClient?.onResume(this)
    }

    private fun fetchUAT() {
        uatLabel.text = getString(R.string.uat_fetching)
        merchant.fetchUAT { error, uat ->
            error?.let {
                uatLabel.text = getString(R.string.uat_display, it.message)
            } ?: run {
                checkoutClient = CheckoutClient(uat!!, this)
                uatLabel.text = getString(R.string.uat_display, uat)
            }
        }
    }

    @Suppress("UNUSED_PARAMETER")
    fun fetchOrderId(view: View) {
        val amount = Amount("USD", "10.00")
        val payee = Payee("ahuang-ppcp-demo-sb1@paypal.com")
        val purchaseUnit = PurchaseUnit(amount, payee)
        val orderRequest = OrderRequest("CAPTURE", listOf(purchaseUnit))

        merchant.fetchOrderId(orderRequest) { error, orderId ->
            error?.let {
                orderIdLabel.text = getString(R.string.error_order_id_not_found)
            } ?: run {
                orderId?.let {
                    this.orderId = it
                    orderIdLabel.text = getString(R.string.order_id_display, this.orderId)
                }
            }
        }
    }

    @Suppress("UNUSED_PARAMETER")
    fun initiateCardCheckout(view: View) {
        progressBar.visibility = View.VISIBLE
        statusLabel.text = getString(R.string.checkout_initiated)

        val cardNumber = cardNumberField.text.replace("\\s+".toRegex(), "")
        val expirationMonth = expDateField.month
        val expirationYear = expDateField.year
        val cvv = cvvField.text.toString()

        val cardBuilder = CardBuilder()
                .cardNumber(cardNumber)
                .expirationMonth(expirationMonth)
                .expirationYear(expirationYear)
                .cvv(cvv)

        checkoutClient?.payWithCard(cardBuilder, orderId, this, this)
    }

    @Suppress("UNUSED_PARAMETER")
    fun initiatePayPalCheckout(view: View?) {
        progressBar.visibility = View.VISIBLE
        checkoutClient?.payWithPayPal(this.orderId, this)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.settings -> {
                this.startActivity(Intent(this, SettingsActivity::class.java))
                true
            }
            R.id.reset -> {
                fetchUAT()
                this.orderId = null
                orderIdLabel.text = getString(R.string.order_id_placeholder)
                statusLabel.text = getString((R.string.checkout_placeholder))
                progressBar.visibility = View.INVISIBLE

                cardNumberField.text = null
                expDateField.text = null
                cvvField.text = null

                super.onOptionsItemSelected(item)
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onPaymentMethodNonceCreated(paymentMethodNonce: PaymentMethodNonce?) {
        val oldStatus = statusLabel.text
        val nonceString = paymentMethodNonce!!.nonce
        val newStatus = "$oldStatus\n$nonceString"
        statusLabel.text = newStatus
    }

    override fun onCheckoutComplete(error: Exception?, result: CheckoutResult?) {
        progressBar.visibility = View.INVISIBLE

        // print error message or order id on success
        var status: String? = null
        error?.let {
            status = getString(R.string.checkout_error, it.message)
        } ?: run {
            if (result is CardCheckoutResult) {
                status = getString(R.string.checkout_success, result.orderId)
            } else if (result is PayPalCheckoutResult) {
                val token = result.token
                val payerId = result.payerId
                status = "Token: $token\nPayer Id: $payerId"
            }
        }
        statusLabel.text = status
    }
}
